wallet-new/src/Cardano/Wallet/Kernel/Restore.hs

```diff
@@ -379,54 +381,51 @@ restoreWalletHistoryAsync wallet rootId prefilter progress start (tgtHash, tgtSl
     startingPoint <- case start of
         Nothing -> withNode $ getFirstGenesisBlockHash genesisHash
         Just sh -> nextHistoricalHash sh >>= maybe (throwM $ RestorationSuccessorNotFound sh) pure
-    restore genesisHash startingPoint NoTimingData
+    restore genesisHash startingPoint 1000
   where
     wId :: WalletId
     wId = WalletIdHdRnd rootId
 
-    -- Process the restoration of the block with the given 'HeaderHash'.
-    restore :: GenesisHash -> HeaderHash -> TimingData -> IO ()
-    restore genesisHash hh timing = do
-
-        -- Updating the average rate every 5 blocks.
-        (rate, timing') <- tickTiming 5 timing
-
-        -- Update each account's historical checkpoints
-        block <- getBlockOrThrow genesisHash hh
+    -- Process the restoration of the next `batchSize` blocks (or up until the
+    -- target hash), starting from the given 'HeaderHash'.
+    restore :: GenesisHash -> HeaderHash -> Int -> IO ()
+    restore genesisHash hh batchSize = do
+        -- Keep track of timing
+        startTime <- getCurrentTime
+
+        -- Processing a batch of blocks, flushing to Acid-State only once
+        (updates, (hh', finished)) <- flip runStrictStateT (hh, False) $ sequence $ replicate batchSize $ do
+            (currentHash, finished) <- get
+            if finished then return Nothing else do
+                if tgtHash == currentHash then
+                    put (currentHash, True)
+                else
+                    lift (nextHistoricalHash currentHash) >>= \case
+                        Nothing  -> throwM (RestorationFinishUnreachable tgtHash currentHash)
+                        Just hh' -> put (hh', False)
+                lift $ getPrefilteredBlockOrThrow genesisHash currentHash
+
+        -- Flush to Acid-State & SQLite
+        k <- getSecurityParameter (wallet ^. walletNode)
+        let (blocks, txMetas, slotIds) = unzip3 (catMaybes updates)
```

Comment:
I think I'd rewrite this code slightly; after discussing this code with @KtorZ , the `catMaybes` here has a dual purpose: any EBBs will be `Nothing`, and once we reach the final block halfway the batch, all the elements after that final block will also be `Nothing`. I find the style a bit confusing, and would probably prefer an explicit recursion style with an accumulator. That said, I don't think it's _wrong_ what's here; but I think it does at least need a comment.